File: E-mart_java/Backend/src/main/java/com/example/Services/SubcategoryService.java
package com.example.Services;

import com.example.Models.Subcategory;
import java.util.List;
import java.util.Optional;

public interface SubcategoryService {
    List<Subcategory> getAllSubcategories();
    
    Optional<Subcategory> getSubcategoryById(int id);
    
    List<Subcategory> getSubcategoriesByCategory(int categoryId);
    
//    Subcategory createSubcategory(Subcategory subcategory);
//    
//    Subcategory updateSubcategory(int id, Subcategory subcategory);
//    
//    void deleteSubcategory(int id);
}

File: E-mart_.Net/Backend/Services/SubCategoryService.cs
using E_mart.Models;
using E_mart.Repositories;
using Microsoft.EntityFrameworkCore;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace E_mart.Services
{
    public class SubCategoryService : ISubCategoryService
    {
        private readonly eMartDbContext _context;

        public SubCategoryService(eMartDbContext context)
        {
            _context = context;
        }

        public async Task<List<Subcategory>> GetAllSubcategories()
        {
            return await _context.Subcategories.ToListAsync();
        }

        public async Task<List<Subcategory>> GetSubcategoriesByCategory(int categoryId)
        {
            return await _context.Subcategories
                .Where(s => s.Categoryid == categoryId)
                .ToListAsync();
        }

        public async Task<Subcategory> GetSubcategoryById(int id)
        {
            return await _context.Subcategories.FindAsync(id);
        }
    }
}
